File: args-python/application/schema.py
from typing import List


class Argument:
    def __init__(self, identifier, help_text):
        self.identifier = identifier
        self.help_text = help_text

    def parse(self, value: str):
        pass

    def default(self):
        pass


class StringArgument(Argument):
    def parse(self, value: str):
        return value

    def default(self):
        return ''


class BooleanArgument(Argument):
    def parse(self, value: str):
        return True

    def default(self):
        return False


class IntegerArgument(Argument):
    def parse(self, value: str):
        return int(value)

    def default(self):
        return 0


class ListArgument(Argument):
    def __init__(self, identifier: str, help_text: str, parser):
        self.identifier = identifier
        self.help_text = help_text
        self._parser = parser

    def parse(self, value: str):
        return list(map(self._parser, value.split(',')))

    def default(self):
        return []


class Schema:
    def __init__(self, arguments: List[Argument]):
        self.args_by_identifier = {arg.identifier: arg for arg in arguments}

    def __getitem__(self, item):
        return self.args_by_identifier[item]

    def __iter__(self):
        self.args_by_identifier.items().__iter__()

    def get_help_text(self):
        help_texts = ["Command usage: "]
        for arg in self.args_by_identifier.values():
            type_string = f"({str(arg.__class__.__name__).replace('Argument', '').lower()})"
            help_texts.append(f"\t-{arg.identifier}\t{type_string.ljust(9, ' ')}\t{arg.help_text}")
        return '\n'.join(help_texts)

    def __eq__(self, other):
        return self.args_by_identifier == other.args_by_identifier

File: args-kotlin/src/main/kotlin/nl/mvdb/args/Schema.kt
package nl.mvdb.args

class Schema {
    private var argumentsByIdentifier: Map<String, AbstractArgument<out Any>>

    constructor(arguments:List<AbstractArgument<out Any>>) {
        this.argumentsByIdentifier = arguments.map { it.identifier to it }.toMap()
    }

    fun get(identifier: String) = this.argumentsByIdentifier[identifier]

    fun getHelpText() = this.argumentsByIdentifier.values
        .map {it.toHelpText()}
        .joinToString(System.lineSeparator(), prefix = "Command usage: ${System.lineSeparator()}")


}

abstract class AbstractArgument<T>(val identifier: String, val helpText: String) {
    abstract fun parse(arg: String): T
    abstract fun default(): T

    fun toHelpText(): String {
        val type_string = this.javaClass.simpleName
            .replace("Argument", "")
            .padStart(9, ' ')

        return "\t-${this.identifier}\t${type_string}\t${this.helpText}"
    }
}

class StringArgument(identifier: String, helpText: String) : AbstractArgument<String>(identifier, helpText) {
    override fun parse(arg: String)= arg
    override fun default() = ""
}

class BooleanArgument(identifier: String, helpText: String) : AbstractArgument<Boolean>(identifier, helpText) {
    override fun parse(arg: String) = true
    override fun default() = false
}

class IntArgument(identifier: String, helpText: String) : AbstractArgument<Int>(identifier, helpText) {
    override fun parse(arg: String) = Integer.valueOf(arg)
    override fun default() = 0
}

class ListArgument<T>(identifier: String, helpText: String, val parser: (String) -> T) : AbstractArgument<List<T>>(identifier, helpText) {
    override fun parse(arg: String) = arg.split(',').map { parser(it) }
    override fun default() = emptyList<T>()
}